### PR TITLE
Added a command line option to start minimized

### PR DESCRIPTION
**Pull request by raibtoffoletto**
The options -m or --start-minimized are working fine when launched from the cmd line.

However I'm failing to start the indicator in case it is the default Gtk.Setting is set to false,
I'm no so experienced with D-Bus, so I haven't be able to implement it.
Hope the added options will help, though!

### CODE REVIEW

File: src/Monitor.vala [hunk 1]
```diff
@@ -4,15 +4,35 @@ namespace Monitor {
         private MainWindow window = null;
         public string[] args;
 
-        construct {
-            application_id = "com.github.stsdc.monitor";
+        private static bool start_minimized = false;
+        private static bool status_minimized = false;
+	    private const GLib.OptionEntry[] cmd_options = {
```
Comment (ryonakano) [inline src/Monitor.vala]:
```suggestion
    private const GLib.OptionEntry[] cmd_options = {
```

File: src/Monitor.vala [hunk 2]
```diff
@@ -4,15 +4,35 @@ namespace Monitor {
...
+		    // --start-minimized
```
Comment (ryonakano) [inline src/Monitor.vala]:
```suggestion
        // --start-minimized
```

File: src/Monitor.vala [hunk 3]
```diff
@@ -4,15 +4,35 @@ namespace Monitor {
...
+		    { "start-minimized", 'm', 0, OptionArg.NONE, ref start_minimized, "Start minimized with wingpanel indicator", null },
```
Comment (ryonakano) [inline src/Monitor.vala]:
```suggestion
            { "start-minimized", 'm', 0, OptionArg.NONE, ref start_minimized, "Start minimized with wingpanel indicator", null },
```

File: src/Monitor.vala [hunk 4]
```diff
@@ -4,15 +4,35 @@ namespace Monitor {
...
+		    // list terminator
```
Comment (ryonakano) [inline src/Monitor.vala]:
```suggestion
            // list terminator
```

File: src/Monitor.vala [hunk 5]
```diff
@@ -4,15 +4,35 @@ namespace Monitor {
...
+		    { null }
```
Comment (ryonakano) [inline src/Monitor.vala]:
```suggestion
            { null }
```

File: src/Monitor.vala [hunk 6]
```diff
@@ -4,15 +4,35 @@ namespace Monitor {
...
+	    };
```
Comment (ryonakano) [inline src/Monitor.vala]:
```suggestion
        };
```

File: src/Monitor.vala [hunk 7]
```diff
@@ -28,7 +48,26 @@ namespace Monitor {
         }
 
         public static int main (string [] args) {
-            var app = new MonitorApp ();
+            // add command line options 
+            try {
+		        var opt_context = new OptionContext ("");
```
Comment (ryonakano) [inline src/Monitor.vala]:
```suggestion
                var opt_context = new OptionContext ("");
```

File: src/Monitor.vala [hunk 8]
```diff
@@ -28,7 +48,26 @@ namespace Monitor {
...
+		            opt_context.set_help_enabled (true);
```
Comment (ryonakano) [inline src/Monitor.vala]:
```suggestion
                opt_context.set_help_enabled (true);
```

File: src/Monitor.vala [hunk 9]
```diff
@@ -28,7 +48,26 @@ namespace Monitor {
...
+			        opt_context.add_main_entries (cmd_options, null);
```
Comment (ryonakano) [inline src/Monitor.vala]:
```suggestion
                opt_context.add_main_entries (cmd_options, null);
```

File: src/Monitor.vala [hunk 10]
```diff
@@ -28,7 +48,26 @@ namespace Monitor {
...
+			        opt_context.parse (ref args);
```
Comment (ryonakano) [inline src/Monitor.vala]:
```suggestion
                opt_context.parse (ref args);
```

File: src/Monitor.vala [hunk 11]
```diff
@@ -28,7 +48,26 @@ namespace Monitor {
...
+            } catch (OptionError e) {
+                print ("error: %s\n", e.message);
+			    print ("Run '%s --help' to see a full list of available command line options.\n", args[0]);
```
Comment (ryonakano) [inline src/Monitor.vala]:
```suggestion
                print ("Run '%s --help' to see a full list of available command line options.\n", args[0]);
```

File: src/Monitor.vala [hunk 12]
```diff
@@ -28,7 +48,26 @@ namespace Monitor {
...
+			    return 0;
```
Comment (ryonakano) [inline src/Monitor.vala]:
```suggestion
                return 0;
```

File: src/Monitor.vala [hunk 13]
```diff
@@ -4,15 +4,35 @@ namespace Monitor {
...
+        private static bool start_minimized = false;
+        private static bool status_minimized = false;
+	    private const GLib.OptionEntry[] cmd_options = {
+		    // --start-minimized
+		    { "start-minimized", 'm', 0, OptionArg.NONE, ref start_minimized, "Start minimized with wingpanel indicator", null },
+		    // list terminator
+		    { null }
+	    };
+
+        //contructor replacement, flags added
```
Comment (ryonakano) [inline src/Monitor.vala]:
```suggestion
        // contructor replacement, flags added
```

**Comment by ryonakano** [comment]
Also, I got many warnings like this:

```
** (com.github.stsdc.monitor:10938): WARNING **: 20:26:52.225: Failed to fetch xid: GDBus.Error:org.freedesktop.DBus.Error.UnknownMethod: No such interface 'org.ayatana.bamf.window' on object at path /org/ayatana/bamf/window/75497479

** (com.github.stsdc.monitor:10938): WARNING **: 20:26:52.225: Failed to fetch type: GDBus.Error:org.freedesktop.DBus.Error.UnknownMethod: No such interface 'org.ayatana.bamf.window' on object at path /org/ayatana/bamf/window/75497479

** (com.github.stsdc.monitor:10938): WARNING **: 20:26:52.225: Failed to fetch monitor: GDBus.Error:org.freedesktop.DBus.Error.UnknownMethod: No such interface 'org.ayatana.bamf.window' on object at path /org/ayatana/bamf/window/75497479

** (com.github.stsdc.monitor:10938): WARNING **: 20:26:52.226: Failed to fetch maximized state: GDBus.Error:org.freedesktop.DBus.Error.UnknownMethod: No such interface 'org.ayatana.bamf.window' on object at path /org/ayatana/bamf/window/75497479

** (com.github.stsdc.monitor:10938): WARNING **: 20:26:52.226: Failed to fetch xid: GDBus.Error:org.freedesktop.DBus.Error.UnknownMethod: No such interface 'org.ayatana.bamf.window' on object at path /org/ayatana/bamf/window/75497479

** (com.github.stsdc.monitor:10938): WARNING **: 20:26:52.226: Failed to fetch type: GDBus.Error:org.freedesktop.DBus.Error.UnknownMethod: No such interface 'org.ayatana.bamf.window' on object at path /org/ayatana/bamf/window/75497479

** (com.github.stsdc.monitor:10938): WARNING **: 20:26:52.226: Failed to fetch path: GDBus.Error:org.freedesktop.DBus.Error.UnknownMethod: No such interface 'org.ayatana.bamf.window' on object at path /org/ayatana/bamf/window/75497479

** (com.github.stsdc.monitor:10938): WARNING **: 20:26:52.228: Unable to fetch children: GDBus.Error:org.freedesktop.DBus.Error.UnknownMethod: No such interface 'org.ayatana.bamf.view' on object at path /org/ayatana/bamf/application/61511822


** (com.github.stsdc.monitor:10938): WARNING **: 20:26:52.228: Failed to fetch path: GDBus.Error:org.freedesktop.DBus.Error.UnknownMethod: No such interface 'org.ayatana.bamf.application' on object at path /org/ayatana/bamf/application/61511822
```

I'm not sure whether this is a correct solution, but this issue seems to be fixed by removing `this.show_all ();` in [`src/MainWindow.vala` at line 65](https://github.com/stsdc/monitor/blob/master/src/MainWindow.vala#L65), and changing `this.show ();` to `this.show_all ();` in [`src/MainWindow.vala` at line 80](https://github.com/stsdc/monitor/blob/master/src/MainWindow.vala#L80).

**Comment by ryonakano** [comment]
> However I'm failing to start the indicator in case it is the default Gtk.Setting is set to false,

What about this code?

```diff
--- a/src/Monitor.vala
+++ b/src/Monitor.vala
@@ -29,7 +29,12 @@ namespace Monitor {
 
                 //start minimized 
                 if (status_minimized) {
-                    window.hide ();
+                    if (window.saved_state.indicator_state) {
+                        window.hide ();
+                    } else {
+                        print ("Error: The app can be started minimized only when wingpanel indicator is enabled, starting normally instead.\n");
+                        window.show_all ();
+                    }
                 } else {
                     window.show_all ();
                 }
```

Maybe doing this makes a too nested conditions in `src/Monitor.vala`, so the next one would be better IMO:

```diff
--- a/src/Monitor.vala
+++ b/src/Monitor.vala
@@ -24,17 +24,23 @@ namespace Monitor {
 
         public override void activate () {
             // only have one window
-            if (get_windows () == null) {
-                window = new MainWindow (this);
+            if (get_windows () != null) {
+                window.present ();
+                return;
+            }
 
-                //start minimized 
-                if (status_minimized) {
+            window = new MainWindow (this);
+
+            //start minimized 
+            if (status_minimized) {
+                if (window.saved_state.indicator_state) {
                     window.hide ();
                 } else {
+                    print ("Error: The app can be started minimized only when wingpanel indicator is enabled, starting normally instead.\n");
                     window.show_all ();
                 }
             } else {
-                window.present ();
+                window.show_all ();
             }
 
             var quit_action = new SimpleAction ("quit", null);
```

**Comment by stsdc** [comment]
Yes, @ryonakano is right.

**Comment by ryonakano** [comment]
Fixes #83

**Comment by raibtoffoletto** [comment]
Thanks a lot for the quick response @ryonakano ! The error message is a quickfix for the indicator problem, but ideally it would be interesting to activate it automatically... I'll look further in the matter.

Regarding the DBus warnings, I didn't figure it out (yet 😄️ ) why they occour, I was hopping some of the more experienced guys like you would give me a light.

About the indents I don't even know how they were messy like that, I've been using emelentary Code with the default 4 spaces... it shouldn't have happened. Not even my gEdit is set to use tabs. I'll manually reset indents in my file just to be sure. Have a few busy days ahead, but I'll work on the solution.

**Comment by ryonakano** [comment]
> The error message is a quickfix for the indicator problem, but ideally it would be interesting to activate it automatically... I'll look further in the matter.

Yeah that would be better! I think you just set `window.saved_state.indicator_state` to true if it is not true:

```diff
--- a/src/Monitor.vala
+++ b/src/Monitor.vala
@@ -29,6 +29,10 @@ namespace Monitor {
 
                 //start minimized 
                 if (status_minimized) {
+                    if (!window.saved_state.indicator_state) {
+                        window.saved_state.indicator_state = true;
+                    }
+
                     window.hide ();
                 } else {
                     window.show_all ();
```

(However, the indicator is not shown soon after the app starts: there is a couple of seconds lag, although this way works.)

File: src/Monitor.vala [hunk 14]
```diff
@@ -4,15 +4,35 @@ namespace Monitor {
         private MainWindow window = null;
         public string[] args;
 
-        construct {
-            application_id = "com.github.stsdc.monitor";
+        private static bool start_minimized = false;
+        private static bool status_minimized = false;
+    private const GLib.OptionEntry[] cmd_options = {
```
Comment (ryonakano) [inline src/Monitor.vala]:
```suggestion
        private const GLib.OptionEntry[] cmd_options = {
```

@raibtoffoletto BTW it seems that I was missing another 4 spaces in my suggestion. Could you fix this again?

**Comment by stsdc** [comment]
`--start-minimized` is nice, but `--start-in-background` wouldn't be better? 🤔
Anyway, this option should be also available in Settings popup (maybe in another PR).

**Comment by raibtoffoletto** [comment]
Ok, I've changed it a bit and it seems to be working now. I was trying to use `dbusserver.indicator_state` instead of the `saved_state.indicator_state`. The later works better.
Thanks :)

There is a delay in the indicator as it is (sometimes) also a delay in displaying the info in the main window (the 'calculating' message), it seems (at least here in my system) that using `show_all ()` before `hide ()` helps a little with the delay... I think it's not that bad. Because this is called in Monitor.vala I've used your fix for the DBus warnings, so everything seems smooth now.
I only get a warning related to `GtkCellAreaBox`, but that seems to be a tree-view problem...

**Comment by ryonakano** [comment]
@raibtoffoletto Thanks for the update!

> (edit) regarding the DBus warnings, I thought they were gone, but no, my mistake. I haven't figured it out... web searches also gave me nothing. The alteration of MainWindow changed nothing for me.

Hmm…the alteration of MainWindow.vala does not show any DBus warnings for me though… :thinking: 
How about removing the `show_all ()` which you added at line 37 and making those changes?

Also, since we decided to call this option `start-in-background`, I think we should update variable names and strings which have old notations.

For example:

```diff
--- a/src/Monitor.vala
+++ b/src/Monitor.vala
@@ -8,7 +8,7 @@ namespace Monitor {
         private static bool status_minimized = false;
         private const GLib.OptionEntry[] cmd_options = {
         // --start-minimized
-            { "start-in-background", 'b', 0, OptionArg.NONE, ref start_minimized, "Start minimized with wingpanel indicator", null },
+            { "start-in-background", 'b', 0, OptionArg.NONE, ref start_in_background, "Start in background with wingpanel indicator", null },
             // list terminator
             { null }
         };
```

**Comment by raibtoffoletto** [comment]
I was still getting the DBus warnings when the settings were off and it would start in background. But you were right, the problem was my duplicated show_all at 37. Now it works fine, my bad!

I think we can live with the delay in starting it. I also cleaned up a bit the my code in `Monitor.vala` there were some unnecessary lines.

File: src/Monitor.vala [hunk 15]
```diff
@@ -28,7 +52,21 @@ namespace Monitor {
         }
 
         public static int main (string [] args) {
-            var app = new MonitorApp ();
+
```
Comment (ryonakano) [inline src/Monitor.vala]: Please no new line after declaring a function. See https://elementary.io/docs/code/reference#whitespace for more details

File: src/Monitor.vala [hunk 16]
```diff
@@ -4,17 +4,41 @@ namespace Monitor {
...
+            }
+
+            window = new MainWindow (this);
+
+            //start in background with indicator
+            if (status_background) {
+                if (!window.saved_state.indicator_state) {
+                    window.saved_state.indicator_state = true;
+                }
+                window.hide ();
```
Comment (ryonakano) [inline src/Monitor.vala:40]: I think we may want to add new line after closing brackets (`}`) here, like this:

```diff
                if (!window.saved_state.indicator_state) {
                    window.saved_state.indicator_state = true;
                }
+
                window.hide ();
```

From https://elementary.io/docs/code/reference#whitespace:

> Lines consisting of closing brackets (`}` or `)`) should be followed by an empty 
line, except when followed by another closing bracket or an `else` statement.

File: src/Monitor.vala [hunk 17]
```diff
@@ -28,7 +52,21 @@ namespace Monitor {
...
+                opt_context.set_help_enabled (true);
+                opt_context.add_main_entries (cmd_options, null);
+                opt_context.parse (ref args);
+            } catch (OptionError e) {
+                print ("Error: %s\n", e.message);
+                print ("Run '%s --help' to see a full list of available command line options.\n\n", args[0]);
+                return 0;
+            }
+
+            MonitorApp app = new MonitorApp (start_in_background);
```
Comment (ryonakano) [inline src/Monitor.vala]: You can simply use `var` like [the current code](https://github.com/stsdc/monitor/blob/0d3edab102814da638dcc8ca0ca1ef1071d3ae60/src/Monitor.vala#L31).

File: src/Monitor.vala (hunk 16 above)
Comment (raibtoffoletto) [reply src/Monitor.vala:40]: I personally think the line there breaks the parent `if` code block in an unflattering way. However, changes made to fit the style ;)

**Comment by raibtoffoletto** [comment]
> I tested this branch again and confirmed it works as expected!! Thank you for your works 

Thanks to you guys for the patience ;) this is an important learning process to me! Hope this version is fine now.

**Pull request merged**